File: backend/app/telemetry/resample.py
"""Shared time-grid resampling, used by every telemetry source.

Different sources sample at different native rates (GoPro GPS ~18Hz, GPMD
accel ~200Hz, a phone/Garmin GPX track often 1Hz) — everything downstream
(lap detection, rendering) wants one uniform per-frame time grid instead.
"""

from __future__ import annotations

import numpy as np
import pandas as pd


def smooth_speed_outliers(speed: pd.Series, window: int = 5, mad_threshold: float = 5.0) -> pd.Series:
    """Rejects implausible GPS speed spikes with a Hampel filter (rolling
    median + median-absolute-deviation threshold).

    Both telemetry sources are exposed to this: a GPS chip's own reported
    speed can glitch during multipath/low-satellite-count, and a
    position-delta-derived speed (external_gpx) is even more sensitive --
    a couple of meters of ordinary position jitter divided by a small time
    delta between fixes produces a wildly implausible instantaneous speed.

    A plain small rolling median (the original approach here) only rejects
    a sample that disagrees with *both* its immediate neighbors -- a burst
    of two or more consecutive bad fixes (e.g. a trackside structure
    blocking sky view for a fraction of a second, common right where a
    go-kart track passes under a bridge/netting) drags the window's median
    along with it and survives untouched. Using a wider window plus an
    explicit outlier test (deviation from the local median, scaled against
    the local median absolute deviation) instead of blanket-replacing every
    sample lets the window stay wide enough to outvote a short burst while
    only actually touching samples that are statistically implausible, so a
    genuine (sustained) acceleration/braking ramp is left alone. This is
    still fundamentally limited by "majority of the window must be good
    fixes" -- a dropout lasting more than about half the window can't be
    distinguished from a real sustained speed change by this alone.
    """
    if len(speed) < window:
        return speed
    median = speed.rolling(window, center=True, min_periods=1).median()
    deviation = (speed - median).abs()
    mad = deviation.rolling(window, center=True, min_periods=1).median()
    # 1.4826 makes MAD comparable to a standard deviation for normally
    # distributed data; floored so a near-constant local window (MAD ~ 0)
    # doesn't make ordinary tiny wobble register as an outlier.
    scaled_mad = (mad * 1.4826).clip(lower=1.0)
    is_outlier = deviation > mad_threshold * scaled_mad
    return speed.where(~is_outlier, median)


def resample_to_grid(df: pd.DataFrame, duration_sec: float, target_fps: float, value_cols: list[str]) -> pd.DataFrame:
    """Interpolates `value_cols` (indexed by df['time']) onto a uniform grid.

    Returns a DataFrame with a 'time' column plus interpolated `value_cols`,
    one row per 1/target_fps step from 0 to duration_sec.
    """
    if df.empty:
        return pd.DataFrame(columns=["time", *value_cols])

    t_target = np.arange(0, duration_sec, 1 / target_fps)
    indexed = df.set_index("time")[value_cols]
    resampled = (
        indexed.reindex(indexed.index.union(t_target))
        .interpolate(method="index")
        # interpolate(method="index") only fills *between* known points --
        # grid steps before the first sample or after the last (e.g. a video
        # trimmed slightly longer than the telemetry's own coverage) are left
        # NaN, which breaks both downstream math and JSON serialization.
        .ffill()
        .bfill()
        .reindex(t_target)
        .reset_index()
        .rename(columns={"index": "time"})
    )
    return resampled

File: frontend/src/lib/telemetry/resample.ts
/**
 * Shared time-grid resampling, used by every telemetry source. Ported from
 * app/telemetry/resample.py -- see that file's docstrings for the reasoning
 * behind the Hampel filter and the per-source-rate window sizing.
 */

/** pandas' `.rolling(window, center=True, min_periods=1)` window bounds:
 * for output index i, the window spans [i - leftOffset, i + rightOffset]
 * clipped to the array, where rightOffset = floor((window-1)/2) and
 * leftOffset = (window-1) - rightOffset. Verified empirically against
 * pandas for both odd and even window sizes (both occur here -- window is
 * derived from a sample-rate ratio and rounded, so it can land on either).
 */
function rollingMedianCentered(values: number[], window: number): number[] {
  const n = values.length
  const rightOffset = Math.floor((window - 1) / 2)
  const leftOffset = window - 1 - rightOffset
  const out = new Array<number>(n)
  for (let i = 0; i < n; i++) {
    const lo = Math.max(0, i - leftOffset)
    const hi = Math.min(n - 1, i + rightOffset)
    const window_ = values.slice(lo, hi + 1).slice().sort((a, b) => a - b)
    const m = window_.length
    out[i] = m % 2 === 1 ? window_[(m - 1) / 2] : (window_[m / 2 - 1] + window_[m / 2]) / 2
  }
  return out
}

/** pandas' `.rolling(window, center=True).mean()` (no `min_periods`
 * override, so it defaults to requiring the *full* window) -- unlike
 * `rollingMedianCentered` above, an index whose clipped window would be
 * short is NaN, not a partial-window average. Used by the GoPro
 * accelerometer smoothing, which fills those edge NaNs with 0 afterwards. */
export function rollingMeanCentered(values: number[], window: number): number[] {
  const n = values.length
  const rightOffset = Math.floor((window - 1) / 2)
  const leftOffset = window - 1 - rightOffset
  const out = new Array<number>(n)
  for (let i = 0; i < n; i++) {
    const lo = i - leftOffset
    const hi = i + rightOffset
    if (lo < 0 || hi > n - 1) {
      out[i] = NaN
      continue
    }
    let sum = 0
    for (let j = lo; j <= hi; j++) sum += values[j]
    out[i] = sum / window
  }
  return out
}

/** Rejects implausible GPS speed spikes with a Hampel filter (rolling
 * median + median-absolute-deviation threshold). See resample.py's
 * `smooth_speed_outliers` docstring for the full rationale. */
export function smoothSpeedOutliers(
  speed: number[],
  window = 5,
  madThreshold = 5.0,
  time?: number[],
  targetWindowSec = 0.5,
): number[] {
  let w = window
  if (time && time.length >= 2) {
    const diffs = time.slice(1).map((t, i) => t - time[i])
    const medianDt = median(diffs)
    if (medianDt > 0) w = Math.max(3, Math.round(targetWindowSec / medianDt))
  }
  if (speed.length < w) return speed.slice()

  const localMedian = rollingMedianCentered(speed, w)
  const deviation = speed.map((v, i) => Math.abs(v - localMedian[i]))
  const mad = rollingMedianCentered(deviation, w)
  // 1.4826 makes MAD comparable to a standard deviation for normally
  // distributed data; floored so a near-constant local window (MAD ~ 0)
  // doesn't make ordinary tiny wobble register as an outlier.
  const scaledMad = mad.map((m) => Math.max(m * 1.4826, 1.0))
  return speed.map((v, i) => (deviation[i] > madThreshold * scaledMad[i] ? localMedian[i] : v))
}

function median(values: number[]): number {
  const sorted = values.slice().sort((a, b) => a - b)
  const m = sorted.length
  if (m === 0) return NaN
  return m % 2 === 1 ? sorted[(m - 1) / 2] : (sorted[m / 2 - 1] + sorted[m / 2]) / 2
}

/** Interpolates `valueCols` (indexed by each row's `time`) onto a uniform
 * grid from 0 to durationSec in steps of 1/targetFps.
 *
 * Equivalent to the Python version's
 * `reindex(union).interpolate(method="index").ffill().bfill().reindex(target)`:
 * for each target time, linearly interpolate (by real time, not position)
 * between the two nearest known samples, clamping to the first/last known
 * value outside the samples' own time range.
 */
export function resampleToGrid<T extends Record<string, number>>(
  rows: (T & { time: number })[],
  durationSec: number,
  targetFps: number,
  valueCols: (keyof T & string)[],
): (Record<string, number> & { time: number })[] {
  const targetTimes: number[] = []
  for (let t = 0; t < durationSec; t += 1 / targetFps) targetTimes.push(t)

  if (rows.length === 0) {
    return targetTimes.map((time) => {
      const row: Record<string, number> = { time }
      for (const col of valueCols) row[col] = NaN
      return row as Record<string, number> & { time: number }
    })
  }

  const sorted = rows.slice().sort((a, b) => a.time - b.time)
  const times = sorted.map((r) => r.time)

  return targetTimes.map((t) => {
    const row: Record<string, number> = { time: t }
    // First index with times[idx] >= t (lower_bound).
    let lo = 0
    let hi = times.length
    while (lo < hi) {
      const mid = (lo + hi) >> 1
      if (times[mid] < t) lo = mid + 1
      else hi = mid
    }
    for (const col of valueCols) {
      if (lo <= 0) {
        row[col] = sorted[0][col]
      } else if (lo >= times.length) {
        row[col] = sorted[times.length - 1][col]
      } else if (times[lo] === t) {
        row[col] = sorted[lo][col]
      } else {
        const t0 = times[lo - 1]
        const t1 = times[lo]
        const v0 = sorted[lo - 1][col]
        const v1 = sorted[lo][col]
        const frac = (t - t0) / (t1 - t0)
        row[col] = v0 + frac * (v1 - v0)
      }
    }
    return row as Record<string, number> & { time: number }
  })
}
